flang/lib/Lower/FIRBuilder.cpp

```diff
@@ -54,6 +54,20 @@ mlir::Value Fortran::lower::FirOpBuilder::createRealConstant(
   return create<mlir::ConstantOp>(loc, realType, getFloatAttr(realType, val));
 }
 
+mlir::Value
+Fortran::lower::FirOpBuilder::createRealZeroConstant(mlir::Location loc,
+                                                     mlir::Type realType) {
+  mlir::Attribute attr;
```

Comment:
Shouldn't this just call `createRealConstant` above?
Why a special method just for `0.0`?

Reply:
At first, I did this in Intrinsics.cpp without creating a helper: 
`auto zero = builder.createRealConstant(loc, resultType, llvm::APFloat(0.)` and it later failed with `error: FloatAttr type doesn't match the type implied by its value.`

So I figured, I have to get the float semantics, and I wanted to use  `kindMap.getFloatSemantics`, but it only works if the type is a `fir::RealType` and we may have `mlir::FloatType`. So that all became a bit ugly to simply create a zero constant. I went for a helper so that this boilerplate is not repeated.

I wanted to make the helper accept a double argument so that it cover more cases than zero, but I got an error telling me that I could not build an APFloat from a double + float semantics. (`APFloat(const fltSemantics &Semantics, T V) = delete;`), which actually make sens because double would only correctly convey Float64 constants, so they cannot be mixed with whatever float semantics.

Hence, I resorted to a simple helper for the zero case (its a very special number after all). MLIR also have specialization like `mlir::OpBuiler::getZeroAttr` for it.

Reply:
OK. Thanks for the background. I agree that this is a sharp corner in MLIR that we keep bumping into. I've discussed it a bit with them as well.